base_layer/wallet/src/transaction_service/service.rs

```diff
@@ -673,6 +694,143 @@ where
         Ok(tx_id)
     }
 
+    /// Sends a one side payment transaction to a recipient
+    /// # Arguments
+    /// 'dest_pubkey': The Comms pubkey of the recipient node
+    /// 'amount': The amount of Tari to send to the recipient
+    /// 'fee_per_gram': The amount of fee per transaction gram to be included in transaction
+    pub async fn send_one_sided_transaction(
+        &mut self,
+        dest_pubkey: CommsPublicKey,
+        amount: MicroTari,
+        fee_per_gram: MicroTari,
+        message: String,
+        transaction_broadcast_join_handles: &mut FuturesUnordered<
+            JoinHandle<Result<u64, TransactionServiceProtocolError>>,
+        >,
+    ) -> Result<TxId, TransactionServiceError>
+    {
+        if self.node_identity.public_key() == &dest_pubkey {
+            warn!(target: LOG_TARGET, "One-sided spend-to-self transactions not supported");
+            return Err(TransactionServiceError::OneSidedTransactionError(
+                "One-sided spend-to-self transactions not supported".to_string(),
+            ));
+        }
+
+        // Prepare sender part of the transaction
+
+        let mut stp = self
+            .output_manager_service
+            .prepare_transaction_to_send(
+                amount,
+                fee_per_gram,
+                None,
+                message.clone(),
+                script!(PushPubKey(Box::new(dest_pubkey.clone()))),
+            )
+            .await?;
+        let tx_id = stp.get_tx_id()?;
+        if !stp.is_single_round_message_ready() {
+            return Err(TransactionServiceError::OneSidedTransactionError(
+                "Sender protocol single round message not ready".to_string(),
+            ));
+        }
+
+        let _ = stp
+            .build_single_round_message()
+            .map_err(|e| TransactionServiceProtocolError::new(tx_id, TransactionServiceError::from(e)))?;
+        if !stp.is_collecting_single_signature() {
+            return Err(TransactionServiceError::OneSidedTransactionError(
+                "Sender protocol single round signature not being collected".to_string(),
+            ));
+        }
+        self.output_manager_service
+            .confirm_pending_transaction(tx_id)
+            .await
+            .map_err(|e| TransactionServiceProtocolError::new(tx_id, TransactionServiceError::from(e)))?;
+
+        // Prepare receiver part of the transaction
+
+        // Diffie-Hellman shared secret `k_Ob * K_Sb = K_Ob * k_Sb` results in a public key, which is converted to
+        // bytes to enable conversion into a private key to be used as the spending key
+        // TODO: Add a standardized Diffie-Hellman method to the tari_crypto library, then come back and use it here
```

Comment:
Is there not one?

Reply:
Not currently no

Reply:
Not this one? https://github.com/tari-project/tari-crypto/blob/b9d434b19c48993d47da3a37b22548d6a3bab265/src/ristretto/ristretto_keys.rs#L247

```
impl DiffieHellmanSharedSecret for RistrettoPublicKey {
```

Reply:
Will try it out

Reply:
Ok, that is halfway there. Used it, but we still need a Diffie-Hellman that will return a private key.